LLVM error when building master on Power VM (OSU)

I'm trying to build a 100% fresh install of latest Julia, but I get the following error
```
    PERL base/version_git.jl.phony
Copying in usr/share/doc/julia/examples
    JULIA usr/lib/julia/inference.ji
LLVM ERROR: Cannot select: 0x1001dd63a60: i64 = addrspacecast[0 -> 10] Constant:i64<0>
  0x1001dd658d0: i64 = Constant<0>
In function: japi1_Typeof_0
Makefile:215: recipe for target '/home/ubuntu/julia/usr/lib/julia/inference.ji' failed
make[1]: *** [/home/ubuntu/julia/usr/lib/julia/inference.ji] Error 1
Makefile:100: recipe for target 'julia-inference' failed
make: *** [julia-inference] Error 2
```
In https://github.com/JuliaLang/julia/pull/22603 I got the same error, but following the suggested method from that thread does not help here. I'm on Power VM @ OSU (Open Source Lab). cc @simonbyrne

--- a/deps/llvm.mk
+++ b/deps/llvm.mk
@@ -464,6 +464,7 @@ $(eval $(call LLVM_PATCH,llvm-D37939-Mem2Reg-Also-handle-memcpy))
 $(eval $(call LLVM_PATCH,llvm-D31524-sovers_4.0)) # Remove for 4.0
 $(eval $(call LLVM_PATCH,llvm-D42262-jumpthreading-not-i1))
 $(eval $(call LLVM_PATCH,llvm-3.9-c_api_nullptr))
+$(eval $(call LLVM_PATCH,llvm-PPC-addrspaces)) # PPC
 ifeq ($(BUILD_LLVM_CLANG),1)
 $(eval $(call LLVM_PATCH,compiler_rt-3.9-glibc_2.25.90)) # Remove for 5.0
 endif
@@ -491,6 +492,7 @@ $(eval $(call LLVM_PATCH,llvm-NVPTX-addrspaces)) # NVPTX
 $(eval $(call LLVM_PATCH,llvm-4.0.0-D37576-NVPTX-sm_70)) # NVPTX, Remove for 6.0
 $(eval $(call LLVM_PATCH,llvm-loadcse-addrspace_4.0))
 $(eval $(call LLVM_PATCH,llvm-D42262-jumpthreading-not-i1))
+$(eval $(call LLVM_PATCH,llvm-PPC-addrspaces)) # PPC
 ifeq ($(BUILD_LLVM_CLANG),1)
 $(eval $(call LLVM_PATCH,compiler_rt-3.9-glibc_2.25.90)) # Remove for 5.0
 endif
@@ -505,6 +507,7 @@ $(eval $(call LLVM_PATCH,llvm-5.0-NVPTX-addrspaces)) # NVPTX
 $(eval $(call LLVM_PATCH,llvm-4.0.0-D37576-NVPTX-sm_70)) # NVPTX, Remove for 6.0
 $(eval $(call LLVM_PATCH,llvm-D38765-gvn_5.0)) # Remove for 6.0
 $(eval $(call LLVM_PATCH,llvm-D42262-jumpthreading-not-i1))
+$(eval $(call LLVM_PATCH,llvm-PPC-addrspaces)) # PPC
 endif # LLVM_VER
 
 # Remove hardcoded OS X requirements in compilter-rt cmake build

--- a/deps/llvm.mk
+++ b/deps/llvm.mk
@@ -465,6 +465,7 @@ $(eval $(call LLVM_PATCH,llvm-D31524-sovers_4.0)) # Remove for 4.0
 $(eval $(call LLVM_PATCH,llvm-D42262-jumpthreading-not-i1))
 $(eval $(call LLVM_PATCH,llvm-3.9-c_api_nullptr))
 $(eval $(call LLVM_PATCH,llvm-PPC-addrspaces)) # PPC
+$(eval $(call LLVM_PATCH,llvm-D30114)) # PPC remove for 5.0
 ifeq ($(BUILD_LLVM_CLANG),1)
 $(eval $(call LLVM_PATCH,compiler_rt-3.9-glibc_2.25.90)) # Remove for 5.0
 endif
@@ -493,6 +494,7 @@ $(eval $(call LLVM_PATCH,llvm-4.0.0-D37576-NVPTX-sm_70)) # NVPTX, Remove for 6.0
 $(eval $(call LLVM_PATCH,llvm-loadcse-addrspace_4.0))
 $(eval $(call LLVM_PATCH,llvm-D42262-jumpthreading-not-i1))
 $(eval $(call LLVM_PATCH,llvm-PPC-addrspaces)) # PPC
+$(eval $(call LLVM_PATCH,llvm-D30114)) # PPC remove for 5.0
 ifeq ($(BUILD_LLVM_CLANG),1)
 $(eval $(call LLVM_PATCH,compiler_rt-3.9-glibc_2.25.90)) # Remove for 5.0
 endif

--- a/deps/patches/llvm-D30114.patch
+++ b/deps/patches/llvm-D30114.patch
@@ -1,0 +1,92 @@
+From 0ee8e563c9b4c77183c4a3b5fbbbb0ed71d3d5bb Mon Sep 17 00:00:00 2001
+From: Eli Friedman <efriedma@codeaurora.org>
+Date: Fri, 24 Feb 2017 20:51:36 +0000
+Subject: [PATCH] [CodeGenPrepare] Make -addr-sink-using-gep work with address
+ spaces.
+
+When we construct addressing modes, we use isNoopAddrSpaceCast to ignore
+addrspacecast instructions. Make sure we insert the correct addrspacecast
+when we reconstruct the addressing mode.
+
+Differential Revision: https://reviews.llvm.org/D30114
+
+
+
+git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@296167 91177308-0d34-0410-b5e6-96231b3b80d8
+---
+ lib/CodeGen/CodeGenPrepare.cpp                           |  8 ++++----
+ test/Transforms/CodeGenPrepare/X86/sink-addrspacecast.ll | 11 ++++++++---
+ 2 files changed, 12 insertions(+), 7 deletions(-)
+
+diff --git a/lib/CodeGen/CodeGenPrepare.cpp b/lib/CodeGen/CodeGenPrepare.cpp
+index 09d8b0274e5..2c9395ca4b2 100644
+--- a/lib/CodeGen/CodeGenPrepare.cpp
++++ b/lib/CodeGen/CodeGenPrepare.cpp
+@@ -4167,7 +4167,7 @@ bool CodeGenPrepare::optimizeMemoryInst(Instruction *MemoryInst, Value *Addr,
+     DEBUG(dbgs() << "CGP: Reusing nonlocal addrmode: " << AddrMode << " for "
+                  << *MemoryInst << "\n");
+     if (SunkAddr->getType() != Addr->getType())
+-      SunkAddr = Builder.CreateBitCast(SunkAddr, Addr->getType());
++      SunkAddr = Builder.CreatePointerCast(SunkAddr, Addr->getType());
+   } else if (AddrSinkUsingGEPs ||
+              (!AddrSinkUsingGEPs.getNumOccurrences() && TM &&
+               SubtargetInfo->useAA())) {
+@@ -4273,7 +4273,7 @@ bool CodeGenPrepare::optimizeMemoryInst(Instruction *MemoryInst, Value *Addr,
+           // We need to add this separately from the scale above to help with
+           // SDAG consecutive load/store merging.
+           if (ResultPtr->getType() != I8PtrTy)
+-            ResultPtr = Builder.CreateBitCast(ResultPtr, I8PtrTy);
++            ResultPtr = Builder.CreatePointerCast(ResultPtr, I8PtrTy);
+           ResultPtr = Builder.CreateGEP(I8Ty, ResultPtr, ResultIndex, "sunkaddr");
+         }
+ 
+@@ -4284,12 +4284,12 @@ bool CodeGenPrepare::optimizeMemoryInst(Instruction *MemoryInst, Value *Addr,
+         SunkAddr = ResultPtr;
+       } else {
+         if (ResultPtr->getType() != I8PtrTy)
+-          ResultPtr = Builder.CreateBitCast(ResultPtr, I8PtrTy);
++          ResultPtr = Builder.CreatePointerCast(ResultPtr, I8PtrTy);
+         SunkAddr = Builder.CreateGEP(I8Ty, ResultPtr, ResultIndex, "sunkaddr");
+       }
+ 
+       if (SunkAddr->getType() != Addr->getType())
+-        SunkAddr = Builder.CreateBitCast(SunkAddr, Addr->getType());
++        SunkAddr = Builder.CreatePointerCast(SunkAddr, Addr->getType());
+     }
+   } else {
+     DEBUG(dbgs() << "CGP: SINKING nonlocal addrmode: " << AddrMode << " for "
+diff --git a/test/Transforms/CodeGenPrepare/X86/sink-addrspacecast.ll b/test/Transforms/CodeGenPrepare/X86/sink-addrspacecast.ll
+index c9f49b5d4f8..eb2c985c70e 100644
+--- a/test/Transforms/CodeGenPrepare/X86/sink-addrspacecast.ll
++++ b/test/Transforms/CodeGenPrepare/X86/sink-addrspacecast.ll
+@@ -1,11 +1,14 @@
+-; RUN: opt -S -codegenprepare < %s | FileCheck %s
++; RUN: opt -S -codegenprepare < %s | FileCheck %s -check-prefix=CHECK -check-prefix=INT
++; RUN: opt -S -codegenprepare -addr-sink-using-gep=true < %s | FileCheck %s -check-prefix=CHECK -check-prefix=GEP
+ 
+ target datalayout =
+ "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128"
+ target triple = "x86_64-unknown-linux-gnu"
+ 
+ ; CHECK-LABEL: @load_cast_gep
+-; CHECK: add i64 %sunkaddr, 40
++; INT: add i64 %sunkaddr, 40
++; GEP: [[CAST:%[0-9]+]] = addrspacecast i64* %base to i8 addrspace(1)*
++; GEP: getelementptr i8, i8 addrspace(1)* [[CAST]], i64 40
+ define void @load_cast_gep(i1 %cond, i64* %base) {
+ entry:
+   %addr = getelementptr inbounds i64, i64* %base, i64 5
+@@ -21,7 +24,9 @@ fallthrough:
+ }
+ 
+ ; CHECK-LABEL: @store_gep_cast
+-; CHECK: add i64 %sunkaddr, 20
++; INT: add i64 %sunkaddr, 20
++; GEP: [[CAST:%[0-9]+]] = addrspacecast i64* %base to i8 addrspace(1)*
++; GEP: getelementptr i8, i8 addrspace(1)* [[CAST]], i64 20
+ define void @store_gep_cast(i1 %cond, i64* %base) {
+ entry:
+   %casted = addrspacecast i64* %base to i32 addrspace(1)*
+-- 
+2.16.2
+

--- a/deps/patches/llvm-PPC-addrspaces.patch
+++ b/deps/patches/llvm-PPC-addrspaces.patch
@@ -1,0 +1,29 @@
+From 15899eaab58e96bb7bbe7a14099674e255656a50 Mon Sep 17 00:00:00 2001
+From: Valentin Churavy <v.churavy@gmail.com>
+Date: Fri, 23 Feb 2018 14:41:20 -0500
+Subject: [PATCH] Make AddrSpaceCast noops on PPC
+
+PPC as AArch64 doesn't have address-spaces so we can drop them in the backend
+---
+ lib/Target/PowerPC/PPCISelLowering.h | 5 +++++
+ 1 file changed, 5 insertions(+)
+
+diff --git a/lib/Target/PowerPC/PPCISelLowering.h b/lib/Target/PowerPC/PPCISelLowering.h
+index e60504507d3..c9b89773968 100644
+--- a/lib/Target/PowerPC/PPCISelLowering.h
++++ b/lib/Target/PowerPC/PPCISelLowering.h
+@@ -761,6 +761,11 @@ namespace llvm {
+       ReuseLoadInfo() : IsInvariant(false), Alignment(0), Ranges(nullptr) {}
+     };
+
++    bool isNoopAddrSpaceCast(unsigned SrcAS, unsigned DestAS) const override {
++      // Addrspacecasts are always noops.
++      return true;
++    }
++
+     bool canReuseLoadAddress(SDValue Op, EVT MemVT, ReuseLoadInfo &RLI,
+                              SelectionDAG &DAG,
+                              ISD::LoadExtType ET = ISD::NON_EXTLOAD) const;
+-- 
+2.16.2
+

--- a/src/jitlayers.cpp
+++ b/src/jitlayers.cpp
@@ -1082,8 +1082,13 @@ void jl_dump_native(const char *bc_fname, const char *unopt_bc_fname, const char
 #else
         Optional<Reloc::Model>(),
 #endif
+#if defined(_CPU_PPC_) || defined(_CPU_PPC64_)
+        // On PPC the small model is limited to 16bit offsets
+        CodeModel::Medium,
+#else
         // Use small model so that we can use signed 32bits offset in the function and GV tables
         CodeModel::Small,
+#endif
         CodeGenOpt::Aggressive // -O3 TODO: respect command -O0 flag?
         ));
 

